File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B08_pca/R/B08_pca.R
# ############################################################################
# B8. PCA
# Tema B08_pca — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript B08_pca.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("B8. PCA\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# B8. PCA
# ----------------------------------------------------------------------------
# QUÉ HACE: tienes muchas variables correlacionadas (redundantes). PCA las
# RESUME en unas pocas variables nuevas (los componentes), que son
# COMBINACIONES LINEALES de las originales y NO están correlacionadas entre
# sí. El primer componente es LA DIRECCIÓN DONDE LOS DATOS MÁS SE DISPERSAN.
#
#   S = W L W'    (descomposición espectral de la covarianza)
#     w_j = autovector j = componente j (los LOADINGS)
#     l_j = autovalor    = varianza en esa dirección
#   Scores (coordenadas nuevas): Z = (X - media) W
#   Varianza explicada del componente j: l_j / suma(l)
#
# LOS COEFICIENTES DE CADA COMPONENTE están en pca$rotation:
#   una COLUMNA por componente, una FILA por variable. Norma 1 cada columna.
#   ⚠️ EL SIGNO ES ARBITRARIO (-w también es autovector). R y Python pueden
#      dar signos opuestos: NO es un error.
#
# CÓMO ELEGIR CUÁNTOS COMPONENTES
#   Scree plot (codo) | varianza acumulada (~90%) | Kaiser (l_j > 1)
#
# PARÁMETROS de prcomp()
#   center = TRUE   siempre
#   scale. = TRUE   usa la matriz de CORRELACIÓN (equivale a estandarizar)
#            FALSE  usa la de COVARIANZA
#            ⚠️ PON TRUE si las variables están en unidades distintas
#   prcomp (SVD, divisor n-1) es MÁS ESTABLE que princomp (espectral, div. n)
#
# ✅ Óptimo entre proyecciones lineales · solución cerrada ·
#    elimina multicolinealidad · proyecta datos nuevos
# ❌ SOLO ESTRUCTURA LINEAL · componentes difíciles de interpretar
#    (mezclan todas las variables) · sensible a escala y outliers ·
#    MAXIMIZAR VARIANZA NO ES LO MISMO QUE MAXIMIZAR INFORMACIÓN ÚTIL
# ============================================================================

pca <- prcomp(X, center=TRUE, scale.=TRUE)     # ⬅ scale.=TRUE casi siempre
print(summary(pca))

varexp <- pca$sdev^2 / sum(pca$sdev^2)
cat(sprintf("\nComponentes para el 90%% de varianza: %d\n",
            which(cumsum(varexp) >= .90)[1]))
cat(sprintf("Criterio de Kaiser (lambda>1): %d componentes\n", sum(pca$sdev^2 > 1)))

cat("\nLOADINGS (coeficientes de cada componente):\n")
print(round(pca$rotation[, 1:min(4, ncol(pca$rotation))], 3))
cat("Lee cada COLUMNA: qué variables pesan más en ese componente.\n")

par(mfrow=c(1,3))
plot(pca$sdev^2, type="b", pch=16, xlab="componente", ylab="varianza (lambda)",
     main="Scree plot (codo)")
abline(h=1, col="red", lty=2)
plot(cumsum(varexp), type="b", pch=16, xlab="nº componentes",
     ylab="varianza acumulada", main="Varianza acumulada")
abline(h=.90, col="red", lty=2)
plot(pca$x[,1], pca$x[,2], pch=16, col=if(!is.null(y)) as.integer(as.factor(y)) else "steelblue",
     xlab=sprintf("PC1 (%.1f%%)", 100*varexp[1]),
     ylab=sprintf("PC2 (%.1f%%)", 100*varexp[2]), main="Datos en 2D")
par(mfrow=c(1,1))

biplot(pca, main="Biplot: observaciones + variables")

cat("\nListo. Revisa la carpeta figuras/\n")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/B08_pca/python/B08_pca.py
"""
B8. PCA (Análisis de Componentes Principales)
Tema B08_pca — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python B08_pca.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("B8. PCA (Análisis de Componentes Principales)")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ B8. PCA ═══  (usa Xs)
pca = PCA(n_components=None).fit(Xs)     # ⬅ None = todos; 0.95 = los que expliquen 95%
Z = pca.transform(Xs)                    # Z = los SCORES (coordenadas nuevas)

print("Varianza explicada por componente:")
for i,(lam,r) in enumerate(zip(pca.explained_variance_, pca.explained_variance_ratio_),1):
    print(f"  PC{i}: lambda={lam:7.4f}  ({r:6.2%})  acumulada={pca.explained_variance_ratio_[:i].sum():6.2%}")

n_90 = int(np.argmax(np.cumsum(pca.explained_variance_ratio_) >= 0.90)) + 1
print(f"\nSe necesitan {n_90} componentes para explicar el 90% de la varianza.")
print(f"Criterio de Kaiser (lambda>1): {int((pca.explained_variance_>1).sum())} componentes")

# ── LOS COEFICIENTES (loadings) ────────────────────────────────────────
loadings = pd.DataFrame(pca.components_.T,
                        columns=[f"PC{i+1}" for i in range(pca.n_components_)],
                        index=nombres_X)
print("\nLoadings (coeficientes de cada componente):")
print(loadings.iloc[:, :min(4, loadings.shape[1])].round(3).to_string())
print("\nLee cada COLUMNA: qué variables pesan más en ese componente.")

fig, ax = plt.subplots(1, 3, figsize=(16, 4.2))
ax[0].plot(range(1,len(pca.explained_variance_)+1), pca.explained_variance_, "o-")
ax[0].axhline(1, color="red", ls="--", label="Kaiser (λ=1)")
ax[0].set_xlabel("componente"); ax[0].set_ylabel("varianza (λ)")
ax[0].set_title("Scree plot (codo)"); ax[0].legend()

ax[1].plot(range(1,len(pca.explained_variance_)+1),
           np.cumsum(pca.explained_variance_ratio_), "o-")
ax[1].axhline(.90, color="red", ls="--", label="90%")
ax[1].set_xlabel("nº de componentes"); ax[1].set_ylabel("varianza acumulada")
ax[1].set_title("Varianza acumulada"); ax[1].legend()

col = y if y is not None else "steelblue"
sc = ax[2].scatter(Z[:,0], Z[:,1], c=col, cmap="viridis", alpha=.7)
ax[2].set_xlabel(f"PC1 ({pca.explained_variance_ratio_[0]:.1%})")
ax[2].set_ylabel(f"PC2 ({pca.explained_variance_ratio_[1]:.1%})")
ax[2].set_title("Datos proyectados en 2D")
plt.tight_layout(); plt.show()

# ── Biplot: observaciones + flechas de variables ───────────────────────
esc = np.abs(Z[:,:2]).max() / np.abs(pca.components_[:2].T).max() * 0.6
plt.scatter(Z[:,0], Z[:,1], alpha=.25, s=25)
for i, var in enumerate(nombres_X):
    plt.arrow(0,0, pca.components_[0,i]*esc, pca.components_[1,i]*esc,
              color="red", head_width=esc*.03, lw=1.5)
    plt.text(pca.components_[0,i]*esc*1.15, pca.components_[1,i]*esc*1.15,
             var, color="red", fontsize=8)
plt.xlabel("PC1"); plt.ylabel("PC2"); plt.title("Biplot: observaciones + variables")
plt.show()

print("\nListo. Revisa la carpeta figuras/")
